### PR TITLE
[OTJ] Implementing "commit crime" mechanic

### PR DESCRIPTION
**Pull request by theelk801**
I'm gonna leave this as a PR for now until we get some more clarification on how exactly this mechanic is worded but I'm pretty sure this is how it's supposed to work. Also needs a test which will be easier to write once we have simpler crime cards

### CODE REVIEW

**Comment by theelk801** [comment]
I think this is good to go now, I just wrote a lot of tests

File: Mage/src/main/java/mage/watchers/common/CommittedCrimeWatcher.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        super.reset();
+        criminals.clear();
+    }
+
+    public static boolean checkCriminality(Game game, Ability source) {
+        return game
+                .getState()
+                .getWatcher(CommittedCrimeWatcher.class)
+                .criminals
+                .contains(source.getSourceId());
```
Comment (JayDi85) [inline Mage/src/main/java/mage/watchers/common/CommittedCrimeWatcher.java:49]: `CommittedCrimeTriggeredAbility::getCriminal` return `controllerId`, so must use `source.getControllerId()` here

Comment (theelk801) [reply Mage/src/main/java/mage/watchers/common/CommittedCrimeWatcher.java:49]: good catch, thanks

File: Mage/src/main/java/mage/abilities/common/CommittedCrimeTriggeredAbility.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            default:
+                return null;
+        }
+        if (controllerId == null || ability == null) {
+            return null;
+        }
+        Set<UUID> opponents = game.getOpponents(controllerId);
+        Set<UUID> targets = ability
+                .getModes()
+                .values()
```
[the remark this answers is not in the document]
Comment (theelk801) [reply Mage/src/main/java/mage/abilities/common/CommittedCrimeTriggeredAbility.java:91]: yeah good call

File: Mage/src/main/java/mage/abilities/common/CommittedCrimeTriggeredAbility.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                || targets
+                .stream()
+                .map(game.getStack()::getStackObject)
+                .filter(Objects::nonNull)
+                .map(Controllable::getControllerId)
+                .anyMatch(opponents::contains)
+                // a card in an opponent's graveyard
+                || targets
+                .stream()
+                .filter(uuid -> Zone.GRAVEYARD.match(game.getState().getZone(uuid)))
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/common/CommittedCrimeTriggeredAbility.java:108]: For info: current wiki page contains talking about any zones, not graveyard only:
https://mtg.fandom.com/wiki/Committing_a_crime
shot_240327_222012

[All scryfall's cards](https://scryfall.com/search?q=oracle%3A%22commit+a+crime%22) contains graveyard only. So maybe wiki is outdated. Or maybe other cards will come. Keep in mind.

**Review by JayDi85 — APPROVED** [review]

File: Mage/src/main/java/mage/abilities/common/CommittedCrimeTriggeredAbility.java (hunk 3 above)
Comment (theelk801) [reply Mage/src/main/java/mage/abilities/common/CommittedCrimeTriggeredAbility.java:108]: I had it that way before but the [mechanics article](https://magic.wizards.com/en/news/feature/outlaws-of-thunder-junction-mechanics) says it's just graveyard, I think there was a miscommunication before.

**Review by xenohedron — APPROVED** [review]
a couple very minor suggestions on documentation. looks good.

File: Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/CommittedCrimeTest.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+import mage.abilities.effects.common.GainLifeEffect;
+import mage.constants.PhaseStep;
+import mage.constants.Zone;
+import org.junit.Test;
+import org.mage.test.serverside.base.CardTestPlayerBase;
+
+/**
+ * @author TheElk801
+ */
+public class CommittedCrimeTest extends CardTestPlayerBase {
```
[the remark this answers is not in the document]
Comment (theelk801) [reply Mage.Tests/src/test/java/org/mage/test/cards/abilities/keywords/CommittedCrimeTest.java:13]: yeah there are a lot of things that could be tested and honestly I just don't want to write any more of them

**Pull request merged**